From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexie Malina <alex@t-ch.net>
Date: Sun, 22 Jun 2025 00:58:40 -0500
Subject: [PATCH] No more natural bats


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index ff5893a7144bc3d740675f6833424c858598b9b4..9826247f81a377f303f90f139b64e77c23b3f9ff 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -310,6 +310,14 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      */
     public static final GameRule<Boolean> LOCATOR_BAR = new GameRule<>("locatorBar", Boolean.class);
 
+
+    // Cuneiform START - No more natural bats.
+    /**
+     * Whether bats will naturally spawn.
+     */
+    public static final GameRule<Boolean> DO_BAT_SPAWN = new GameRule<>("doBatSpawn", Boolean.class);
+    // Cuneiform END
+
     // All GameRules instantiated above this for organizational purposes
     private final String name;
     private final Class<T> type;
