From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexie Malina <alex@t-ch.net>
Date: Mon, 23 Jun 2025 14:25:59 -0500
Subject: [PATCH] Mob Gamerules


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index 9826247f81a377f303f90f139b64e77c23b3f9ff..657a5c497c9fedcbfbc77bf6c0081c7b5fc33d5e 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -311,11 +311,36 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
     public static final GameRule<Boolean> LOCATOR_BAR = new GameRule<>("locatorBar", Boolean.class);
 
 
-    // Cuneiform START - No more natural bats.
+    // Cuneiform START - GameRules
     /**
      * Whether bats will naturally spawn.
      */
     public static final GameRule<Boolean> DO_BAT_SPAWN = new GameRule<>("doBatSpawn", Boolean.class);
+
+    /**
+     * Whether the wither will destroy blocks.
+     */
+    public static final GameRule<Boolean> WITHER_GRIEFING = new GameRule<>("witherGriefing", Boolean.class);
+
+    /**
+     * Whether creepers will destroy blocks.
+     */
+    public static final GameRule<Boolean> CREEPER_GRIEFING = new GameRule<>("creeperGriefing", Boolean.class);
+
+    /**
+     * Whether endermen will carry blocks.
+     */
+    public static final GameRule<Boolean> ENDERMAN_GRIEFING = new GameRule<>("endermanGriefing", Boolean.class);
+
+    /**
+     * Whether ghasts can explode blocks,
+     */
+    public static final GameRule<Boolean> GHAST_GRIEFING = new GameRule<>("ghastGriefing", Boolean.class);
+
+    /**
+     * Whether mobs that use BreakDoorGoal will break doors.
+     */
+    public static final GameRule<Boolean> BREAK_DOORS = new GameRule<>("breakDoors", Boolean.class);
     // Cuneiform END
 
     // All GameRules instantiated above this for organizational purposes
