From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexie Malina <alex@t-ch.net>
Date: Mon, 23 Jun 2025 14:26:41 -0500
Subject: [PATCH] noXPCooldown GameRule


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index 657a5c497c9fedcbfbc77bf6c0081c7b5fc33d5e..f388664dc57308dcc82880cf72b14681dc21a172 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -341,6 +341,11 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      * Whether mobs that use BreakDoorGoal will break doors.
      */
     public static final GameRule<Boolean> BREAK_DOORS = new GameRule<>("breakDoors", Boolean.class);
+
+    /**
+     * Whether the collection of experience points should have a cooldown.
+     */
+    public static final GameRule<Boolean> NO_XP_COOLDOWN = new GameRule<>("noXPCooldown", Boolean.class);
     // Cuneiform END
 
     // All GameRules instantiated above this for organizational purposes
